File: backend/recent_store.py
"""SQLite-backed recent-searches store.

Single table; one row per search event. Queries dedupe by PN at read time
so the Home screen shows each part once (most recent lookup wins).
"""
from __future__ import annotations

import os
import sqlite3
import threading
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

_DB_PATH = Path(
    os.getenv("RECENT_DB_PATH", "./recent_searches.db")
).resolve()
_lock = threading.Lock()
_initialized = False


def _connect() -> sqlite3.Connection:
    # Allow the same connection to be shared across threads when Flask runs
    # in multi-threaded mode (the module lock serialises writes).
    conn = sqlite3.connect(str(_DB_PATH), check_same_thread=False, timeout=5.0)
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA journal_mode=WAL")
    return conn


def init() -> None:
    global _initialized  # noqa: PLW0603
    with _lock:
        if _initialized:
            return
        _DB_PATH.parent.mkdir(parents=True, exist_ok=True)
        with _connect() as conn:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS recent_searches (
                    id            INTEGER PRIMARY KEY AUTOINCREMENT,
                    pn            TEXT    NOT NULL,
                    mpn           TEXT    DEFAULT '',
                    manufacturer  TEXT    DEFAULT '',
                    com_id        TEXT    DEFAULT '',
                    lifecycle     TEXT    DEFAULT '',
                    yeol          REAL,
                    risk          REAL,
                    source        TEXT    DEFAULT '',
                    kind          TEXT    DEFAULT 'single',
                    searched_at   TEXT    NOT NULL
                )
                """
            )
            conn.execute(
                "CREATE INDEX IF NOT EXISTS idx_recent_pn ON recent_searches(pn)"
            )
            conn.execute(
                "CREATE INDEX IF NOT EXISTS idx_recent_time "
                "ON recent_searches(searched_at DESC)"
            )
            conn.commit()
        _initialized = True


def record(
    *,
    pn: str,
    mpn: str = "",
    manufacturer: str = "",
    com_id: str = "",
    lifecycle: str = "",
    yeol: float | None = None,
    risk: float | None = None,
    source: str = "",
    kind: str = "single",
) -> None:
    """Append a single search event. Fail-soft: never raises into the request."""
    try:
        init()
        pn = (pn or "").strip()
        if not pn:
            return
        now = datetime.now(timezone.utc).isoformat(timespec="seconds")
        with _lock, _connect() as conn:
            conn.execute(
                """
                INSERT INTO recent_searches
                    (pn, mpn, manufacturer, com_id, lifecycle, yeol, risk,
                     source, kind, searched_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    pn,
                    mpn or "",
                    manufacturer or "",
                    com_id or "",
                    lifecycle or "",
                    yeol,
                    risk,
                    source or "",
                    kind or "single",
                    now,
                ),
            )
            conn.commit()
    except Exception:  # noqa: BLE001 — telemetry must never break the user's request
        pass


def list_recent(limit: int = 10) -> list[dict[str, Any]]:
    """Return the most recent unique PNs (limit rows). Newest first."""
    init()
    with _connect() as conn:
        # Window function would be ideal; stdlib sqlite3 supports it since 3.25.
        rows = conn.execute(
            """
            SELECT pn, mpn, manufacturer, com_id, lifecycle, yeol, risk,
                   source, kind, searched_at
            FROM (
                SELECT *,
                       ROW_NUMBER() OVER (PARTITION BY pn
                                          ORDER BY searched_at DESC) AS rn
                FROM recent_searches
            )
            WHERE rn = 1
            ORDER BY searched_at DESC
            LIMIT ?
            """,
            (int(limit),),
        ).fetchall()
    return [
        {
            "pn":            r["pn"],
            "mpn":           r["mpn"] or "",
            "manufacturer":  r["manufacturer"] or "",
            "comId":         r["com_id"] or "",
            "lifecycle":     r["lifecycle"] or "",
            "yeol":          r["yeol"],
            "risk":          r["risk"],
            "source":        r["source"] or "",
            "kind":          r["kind"] or "single",
            "searchedAt":    r["searched_at"],
        }
        for r in rows
    ]


def clear() -> int:
    init()
    with _lock, _connect() as conn:
        n = conn.execute("DELETE FROM recent_searches").rowcount
        conn.commit()
    return n

File: lib/recentStore.ts
/**
 * File-backed recent-searches store for the Next.js runtime.
 *
 * Stdlib-only, no binary dependencies. Reads/writes a JSON file at
 * RECENT_DB_PATH (default ./recent_searches.json) with an advisory
 * per-process mutex around writes. Matches the shape served by the
 * Flask backend's /api/recent so the frontend code is identical.
 */
import fs from 'node:fs';
import path from 'node:path';

export type RecentEntry = {
  pn: string;
  mpn: string;
  manufacturer: string;
  comId: string;
  lifecycle: string;
  yeol: number | null;
  risk: number | null;
  source: string;
  kind: string;
  searchedAt: string;
};

const FILE = path.resolve(
  process.env.RECENT_DB_PATH?.replace(/\.db$/, '.json') ?? './recent_searches.json',
);

let writing: Promise<void> = Promise.resolve();

function readAll(): RecentEntry[] {
  try {
    const text = fs.readFileSync(FILE, 'utf8');
    const data = JSON.parse(text);
    return Array.isArray(data) ? (data as RecentEntry[]) : [];
  } catch {
    return [];
  }
}

function writeAll(rows: RecentEntry[]): void {
  try {
    fs.mkdirSync(path.dirname(FILE), { recursive: true });
    fs.writeFileSync(FILE, JSON.stringify(rows, null, 2));
  } catch {
    /* fail-soft — telemetry must never break a request */
  }
}

export async function record(entry: Partial<RecentEntry> & { pn: string }): Promise<void> {
  const pn = (entry.pn || '').trim();
  if (!pn) return;
  await (writing = writing.then(() => {
    const rows = readAll();
    rows.push({
      pn,
      mpn: entry.mpn ?? '',
      manufacturer: entry.manufacturer ?? '',
      comId: entry.comId ?? '',
      lifecycle: entry.lifecycle ?? '',
      yeol: entry.yeol ?? null,
      risk: entry.risk ?? null,
      source: entry.source ?? '',
      kind: entry.kind ?? 'single',
      searchedAt: new Date().toISOString(),
    });
    // Cap storage at a sensible number so the file stays small.
    if (rows.length > 5000) rows.splice(0, rows.length - 5000);
    writeAll(rows);
  }).catch(() => undefined));
}

export function listRecent(limit = 10): RecentEntry[] {
  const rows = readAll();
  // Dedupe by PN, keep most recent per PN, newest first.
  const byPN = new Map<string, RecentEntry>();
  for (const r of rows) {
    const prev = byPN.get(r.pn);
    if (!prev || r.searchedAt > prev.searchedAt) byPN.set(r.pn, r);
  }
  return [...byPN.values()]
    .sort((a, b) => (a.searchedAt < b.searchedAt ? 1 : -1))
    .slice(0, limit);
}

export async function clear(): Promise<number> {
  const n = readAll().length;
  writeAll([]);
  return n;
}
